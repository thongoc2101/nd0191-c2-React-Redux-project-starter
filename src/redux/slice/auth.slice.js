import { createSlice } from "@reduxjs/toolkit";
import { AuthActions } from "../actions/auth.action";

const initialState = {
    userInfo: null,
    users: null,
    isLogin: false
};

export const AuthSlice = createSlice({
    name: 'auth',
    initialState: initialState,
    reducers: {
        login: (state, action) => {
            const user = Object.values(state.users).find(user => user.id === action.payload.username && user.password === action.payload.password);
            if (user) {
                state.isLogin = true;
                state.userInfo = user;
            } else {
                state.isLogin = false;
                state.userInfo = null;
            }
        },
        logout: (state) => {
            state.isLogin = false;
            state.userInfo = null;
        },

        setQuestionUser: (state, action) => {
            const userItem = Object.values(state.users).find(user => user.id === action.payload.author);
            userItem.questions.push(action.payload.id);

            state.userInfo = userItem;
            state.users = {
                ...state.users,
                [userItem.id]: userItem
            }
        },

        setUserAnswer: (state, action) => {
            const authedUser = action.payload.authedUser;
            const answer = action.payload.answer;
            const qid = action.payload.id;

            state.users = {
                ...state.users,
                [authedUser]: {
                  ...state.users[authedUser],
                  answers: {
                    ...state.users[authedUser].answers,
                    [qid]: answer
                  }
                }
              }
        }
    },
    extraReducers: (builder) => {
        // Add extra reducers here
        builder.addCase(AuthActions.getUsers.fulfilled, (state, action) => {
            state.users = action.payload;
        });
    }
});

export const { login, logout, setQuestionUser, setUserAnswer } = AuthSlice.actions;