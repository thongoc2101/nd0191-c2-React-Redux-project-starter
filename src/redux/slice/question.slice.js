import { createSlice } from "@reduxjs/toolkit";
import { QuestionActions } from "../actions/question.action";

const initialState = {
    questions: null,
};

export const QuestionSlice = createSlice({
    name: 'questions',
    initialState: initialState,
    reducers: {
        setAnsweredUser: (state, action) => {
            const qid = action.payload.id;
            const answer = action.payload.answer;
            const authedUser = action.payload.authedUser;

            state.questions = {
                ...state.questions,
                [qid]: {
                  ...state.questions[qid],
                  [answer]: {
                    ...state.questions[qid][answer],
                    votes: state.questions[qid][answer].votes.push(authedUser)
                  }
                }
              }
        }
    },
    extraReducers: (builder) => {
        // Add extra reducers here
        builder.addCase(QuestionActions.getQuestions.fulfilled, (state, action) => {
            state.questions = action.payload;
        });
    }
});

export const { setAnsweredUser } = QuestionSlice.actions;