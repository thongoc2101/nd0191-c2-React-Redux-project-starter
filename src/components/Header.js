import { useDispatch, useSelector } from "react-redux";
import { Link, useNavigate } from "react-router-dom";
import { logout } from "../redux/slice/auth.slice";

const Header = () => {
    const dispatch = useDispatch();
    const navigate = useNavigate();
    const { userInfo } = useSelector((state) => state.auth);

    const handleLogout = () => {
        dispatch(logout());
        navigate('/login');
    }

    return (
        <nav className="grid grid-cols-2 gap-4">
            <div className="left-information">
                <Link to="/" className="font-medium px-3 py-2 text-slate-700 rounded-md">Home</Link>
                <Link to="/leaderboard" className="font-medium px-3 py-2 text-slate-700 rounded-md">Leaderboard</Link>
                <Link to="/add" className="font-medium px-3 py-2 text-slate-700 rounded-md">New</Link>
            </div>

            <div className="right-information">
                <span
                    className="font-medium px-3 py-2 text-slate-700">{userInfo?.name}</span>
                <button onClick={handleLogout}
                    className="font-medium px-3 py-2 text-slate-700 rounded-md">Logout
                </button>
            </div>
        </nav>
    );
}

export default Header;