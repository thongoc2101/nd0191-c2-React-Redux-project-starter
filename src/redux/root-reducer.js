import { combineReducers } from "@reduxjs/toolkit";
import { AuthSlice } from "./slice/auth.slice";
import { QuestionSlice } from "./slice/question.slice";

export const rootReducer = combineReducers({
    auth: AuthSlice.reducer,
    questions: QuestionSlice.reducer,
});