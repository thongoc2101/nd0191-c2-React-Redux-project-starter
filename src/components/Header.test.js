import { render } from '@testing-library/react';
import { Provider } from 'react-redux';
import { BrowserRouter } from 'react-router-dom';
import {store} from "../redux/store";
import Header from './Header';
import { AuthActions } from '../redux/actions/auth.action';
import { login } from '../redux/slice/auth.slice';

describe("Header", () => {
    it("should render the component when logged in", async () => {
        await store.dispatch(AuthActions.getUsers());
        await store.dispatch(login({ username: 'tylermcginnis', password: 'abc321'}))

        const component = render(
            <Provider store={store}>
                <BrowserRouter>
                    <Header />
                </BrowserRouter>
            </Provider>
        );

        expect(component).toBeDefined();
        expect(component).toMatchSnapshot();
    });
});
