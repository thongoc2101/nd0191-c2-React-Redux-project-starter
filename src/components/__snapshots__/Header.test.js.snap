// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header should render the component when logged in 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <nav
        class="grid grid-cols-2 gap-4"
      >
        <div
          class="left-information"
        >
          <a
            class="font-medium px-3 py-2 text-slate-700 rounded-md"
            href="/"
          >
            Home
          </a>
          <a
            class="font-medium px-3 py-2 text-slate-700 rounded-md"
            href="/leaderboard"
          >
            Leaderboard
          </a>
          <a
            class="font-medium px-3 py-2 text-slate-700 rounded-md"
            href="/add"
          >
            New
          </a>
        </div>
        <div
          class="right-information"
        >
          <span
            class="font-medium px-3 py-2 text-slate-700"
          >
            Tyler McGinnis
          </span>
          <button
            class="font-medium px-3 py-2 text-slate-700 rounded-md"
          >
            Logout
          </button>
        </div>
      </nav>
    </div>
  </body>,
  "container": <div>
    <nav
      class="grid grid-cols-2 gap-4"
    >
      <div
        class="left-information"
      >
        <a
          class="font-medium px-3 py-2 text-slate-700 rounded-md"
          href="/"
        >
          Home
        </a>
        <a
          class="font-medium px-3 py-2 text-slate-700 rounded-md"
          href="/leaderboard"
        >
          Leaderboard
        </a>
        <a
          class="font-medium px-3 py-2 text-slate-700 rounded-md"
          href="/add"
        >
          New
        </a>
      </div>
      <div
        class="right-information"
      >
        <span
          class="font-medium px-3 py-2 text-slate-700"
        >
          Tyler McGinnis
        </span>
        <button
          class="font-medium px-3 py-2 text-slate-700 rounded-md"
        >
          Logout
        </button>
      </div>
    </nav>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
